### PR TITLE
In avoid to call zk frequently, using watcher to watch the change

### LINKED ISSUE #3358: ZkLedgerUnderreplicationManager: cache responses / reduce number of zk calls by isLedgerReplicationEnabled()
It looks like ZkLedgerUnderreplicationManager.isLedgerReplicationEnabled() can cache the state and either use watcher (see notifyLedgerReplicationEnabled) + enableLedgerReplication/disableLedgerReplication or just some time interval (10sec or so) to cache the value for.

AFAICT Auditor.checkAllLedgers() calls it rather frequently

### PR DESCRIPTION
**Pull request by gaozhangmin**
Fixes #3358

### CODE REVIEW

**Comment by gaozhangmin** [comment]
@dlg99  PTAL Thx

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManager.java [hunk 1]
```diff
@@ -737,6 +737,40 @@ public boolean isLedgerReplicationEnabled()
...
+                    cb.operationComplete(0, null);
+                } else if (e.getType() == Watcher.Event.EventType.NodeCreated) {
+                    LOG.info("bLedgerReplication is disabled externally through Zookeeper, "
+                            + "since DISABLE_NODE ZNode is created");
+                    cb.operationComplete(1, null);
+                }
+            }
+        };
+        try {
+            zkc.addWatch(basePath + '/'
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManager.java:761]: We must implement this for every supported Metadata driver 
What happens if the driver doesn't implement it?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java [hunk 2]
```diff
@@ -167,6 +167,7 @@ public class Auditor implements AutoCloseable {
     private final int zkOpTimeoutMs;
     private final Semaphore openLedgerNoRecoverySemaphore;
     private final int openLedgerNoRecoverySemaphoreWaitTimeoutMSec;
+    private boolean isLedgerReplicationEnabled;
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:170]: This should be a AtomicBoolean as it is accessed by multiple threads

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManager.java (hunk 1 above)
Comment (gaozhangmin) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManager.java:761]: If this, We also have this problem  currently.  see `notifyLedgerReplicationEnabled `

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManager.java [hunk 3]
```diff
@@ -737,6 +737,40 @@ public boolean isLedgerReplicationEnabled()
         }
     }
 
+    @Override
+    public void notifyLedgerReplicationStatusChanged(final GenericCallback<Void> cb)
+            throws ReplicationException.UnavailableException {
```
Comment (dlg99) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManager.java:742]: ReplicationException.fromKeeperException can throw NonRecoverableReplicationException

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java [hunk 4]
```diff
@@ -818,11 +821,6 @@ public void run() {
                         LOG.error("Exception running periodic check", bke);
                     } catch (IOException ioe) {
                         LOG.error("I/O exception running periodic check", ioe);
-                    } catch (ReplicationException.NonRecoverableReplicationException nre) {
-                        LOG.error("Non Recoverable Exception while reading from ZK", nre);
-                        submitShutdownTask();
```
Comment (dlg99) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:823]: why is this removed / where NonRecoverableReplicationException is handled now?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java [hunk 5]
```diff
@@ -1328,18 +1326,8 @@ void checkAllLedgers() throws BKException, IOException, InterruptedException {
...
-                try {
-                    if (!ledgerUnderreplicationManager.isLedgerReplicationEnabled()) {
-                        LOG.info("Ledger rereplication has been disabled, aborting periodic check");
-                        FutureUtils.complete(processFuture, null);
-                        return;
-                    }
-                } catch (ReplicationException.NonRecoverableReplicationException nre) {
-                    LOG.error("Non Recoverable Exception while reading from ZK", nre);
-                    submitShutdownTask();
-                    return;
```
Comment (dlg99) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:1340]: why is this removed / where NonRecoverableReplicationException is handled now?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerUnderreplicationManager.java [hunk 6]
```diff
@@ -148,6 +148,16 @@ void enableLedgerReplication()
...
             throws ReplicationException.UnavailableException;
 
+    /**
+     * Receive notification asynchronously when the ledger replication process
+     * is changed.
+     *
+     * @param cb
+     *            - callback implementation to receive the notification
+     */
+    void notifyLedgerReplicationStatusChanged(GenericCallback<Void> cb)
```
Comment (dlg99) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerUnderreplicationManager.java:158]: something like
``` java
default void notifyLedgerReplicationStatusChanged(GenericCallback<Void> cb) throws ReplicationException.UnavailableException {
 throw new RuntimeException("not implemented");
}
```

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java (hunk 2 above)
Comment (dlg99) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:170]: +1

**Review by dlg99 — CHANGES_REQUESTED** [review]

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java (hunk 5 above)
Comment (gaozhangmin) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:1346]: it's handled where `notifyLedgerReplicationStatusChanged` is called

**Comment by gaozhangmin** [comment]
@eolivelli  @dlg99  PTAL

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerUnderreplicationManager.java [hunk 7]
```diff
@@ -148,6 +148,18 @@ void enableLedgerReplication()
...
+    /**
+     * Receive notification asynchronously when the ledger replication process
+     * is changed.
+     *
+     * @param cb
+     *            - callback implementation to receive the notification
+     */
+    default void notifyLedgerReplicationStatusChanged(GenericCallback<Void> cb)
+            throws ReplicationException.UnavailableException {
+        throw new RuntimeException("not implemented");
```
Comment (merlimat) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerUnderreplicationManager.java:160]: This would be breaking the (runtime) compatibility for an existing `LedgerUnderReplicationManager` that does not implement the method, since it will throw exception.

**Review by merlimat — COMMENTED** [review]
I think the repeated calls to ZK are not be a problem in Pulsar, since 2.10, because the calls to ZK are already internally cached.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerUnderreplicationManager.java (hunk 7 above)
Comment (gaozhangmin) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerUnderreplicationManager.java]: Remove default  method as current now.

**Comment by gaozhangmin** [comment]
rerun failure checks
